file_system.go

```diff
@@ -125,16 +146,27 @@ func commandNotSupporterdError(err error) bool {
 // be used. You may have to set ServerLocation in your config to get (more)
 // accurate ModTimes in this case.
 func (c *Client) ReadDir(path string) ([]os.FileInfo, error) {
-	entries, err := c.dataStringList("MLSD %s", path)
+	pconn, err := c.getIdleConn()
+	if err != nil {
+		return nil, err
+	}
 
+	defer c.returnConn(pconn)
+
+	var entries []string
+
+	if !pconn.mlsdNotSupported {
+		entries, err = c.dataStringList(pconn, "MLSD %s", path)
+	}
 	parser := parseMLST
 
-	if err != nil {
-		if !commandNotSupporterdError(err) {
+	if pconn.mlsdNotSupported || err != nil {
+		if !pconn.mlsdNotSupported && !commandNotSupporterdError(err) {
```

Comment:
Can this commandNotSupported check be moved up directly after the dataStringList() call? It's a little weird having the nested checks if mlsdNotSupported.